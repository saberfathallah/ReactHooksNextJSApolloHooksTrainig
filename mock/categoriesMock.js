export const CATEGORY = {
    id: "5e89c5f78245cd0d75dec50a",
    level: 1,
    name: "Sport",
    children: [
      {
        children: [],
        id: "5e89c6068245cd0d75dec50c",
        name: "foot",
        level: 2,
      },
    ],
  }