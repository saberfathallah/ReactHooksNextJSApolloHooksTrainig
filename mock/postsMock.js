export const POST = {
  categoryId: "5e89c6458245cd0d75dec50f",
  postId: "5e89c6458245cd0d75dec50f",
  comments: [],
  creatorId: "5e89c2fd8245cd0d75dec501",
  description: "corona france post",
  likes: ["5e8a26784c6b1d234b4c1a3f", "5e89c2fd8245cd0d75dec501"],
  id: "5e89c6878245cd0d75dec510",
  userId: {
    id: "5fac3671dd9b1b079249d522",
    email: "sberrr@gmail.com",
    name: "sberrr",
  },
};

// interface Post {
//   userId: User;
//   categoryId: string;
//   postId: string;
//   comments: Comment[];
//   description: string;
//   id: string;
//   likes: string[];
// }

// categoryId: "5e89c6458245cd0d75dec50f"
// comments: (5) [{…}, {…}, {…}, {…}, {…}]
// description: "corona france post"
// id: "5e89c6878245cd0d75dec510"
// likes: (2) ["5e8a26784c6b1d234b4c1a3f", "5e89c2fd8245cd0d75dec501"]
// userId: {name: "Saber Fathallah", id: "5e89c2fd8245cd0d75dec501", __typename: "User"}
