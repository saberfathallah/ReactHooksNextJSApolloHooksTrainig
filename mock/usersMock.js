export const USER_CONNECTED = {
    id: "5fac3671dd9b1b079249d522",
    email: "sberrr@gmail.com",
    name: "sberrr",
  };