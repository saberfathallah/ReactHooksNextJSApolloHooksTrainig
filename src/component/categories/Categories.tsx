import React from 'react';
import { useQuery, useMutation } from '@apollo/react-hooks';
import { makeStyles } from '@material-ui/core/styles';
import TreeView from '@material-ui/lab/TreeView';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import ChevronRightIcon from '@material-ui/icons/ChevronRight';
import TreeItem from '@material-ui/lab/TreeItem';

import CATEGORIES from '@graphql/categories/queries/getAllCategories';
import CHANGE_CURRENT_CATEGORY from '@graphql/client/mutation/changeCurrentCategory';

interface CategoriesResponse {
  getAllCategoriesQuery: Categories;
}

interface Categories {
  categories: Category[];
}

interface Category {
  level: number;
  name: string;
  id: string;
  children: Category[];
}

const useStyles = makeStyles({
  root: {
    height: 216,
    flexGrow: 1,
    maxWidth: 400,
  },
});

const categoryChildren: any = (
  children, changeCurrentCategory, isDisplayCategories, selectCategoryId,
): Category[] => (children && children.length > 0
  ? children.map((child) => (
    <div key={child.id}>
      <TreeItem
        onClick={(): void => {
          if (isDisplayCategories) {
            changeCurrentCategory({ variables: { currentCategoryId: child.id } });
          } else {
            selectCategoryId(child.id);
          }
        }}
        nodeId={child.id}
        label={child.name}
      >
        {categoryChildren(
          child.children, changeCurrentCategory, isDisplayCategories, selectCategoryId,
        )}
      </TreeItem>
    </div>
  ))
  : null);

const Categories: React.FC<any> = (props: any) => {
  const { selectCategoryId, isDisplayCategories } = props;
  const classes = useStyles();
  const { data, loading } = useQuery<CategoriesResponse>(CATEGORIES);
  const [changeCurrentCategory] = useMutation(CHANGE_CURRENT_CATEGORY);

  if (loading) return <p>loading...</p>;

  return (
    <div>
      <TreeView
        className={classes.root}
        defaultCollapseIcon={<ExpandMoreIcon />}
        defaultExpandIcon={<ChevronRightIcon />}
        multiSelect
      >
        {
          data.getAllCategoriesQuery.categories.map((category) => (
            <div key={category.id}>
              <TreeItem
                onClick={(): void => {
                  if (isDisplayCategories) {
                    changeCurrentCategory({ variables: { currentCategoryId: category.id } });
                  } else {
                    selectCategoryId(category.id);
                  }
                }}
                nodeId={category.id}
                label={category.name}
              >
                {categoryChildren(
                  category.children,
                  changeCurrentCategory,
                  isDisplayCategories,
                  selectCategoryId,
                )}
              </TreeItem>
            </div>
          ))
        }
      </TreeView>
    </div>
  );
};

export default Categories;
