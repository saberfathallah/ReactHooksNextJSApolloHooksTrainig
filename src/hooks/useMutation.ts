import { useMutation as useMutationApollo } from "@apollo/react-hooks";

export const useMutation = (
  mutation: any,
  { variables, optionCache, ...otherOptions }: any
): any => {
  if (optionCache) {
    const [mutationFunction, { data, error, loading }] = useMutationApollo(
      mutation,
      {
        variables,
        update: (cache, { data }) =>
          optionCache.updateCacheFunction(cache, data, optionCache.variables),
          ...otherOptions,
      }
    );
    return [mutationFunction, { data, error, loading }];
  }

  const [mutationFunction, { data, error, loading }] = useMutationApollo(
    mutation,
    {
      variables,
    }
  );

  return [mutationFunction, { data, error, loading }];
};
