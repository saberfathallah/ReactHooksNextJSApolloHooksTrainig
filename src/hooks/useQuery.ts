import { useQuery as useQueryApollo } from "@apollo/react-hooks";

export const useQuery = (query, options={}): any => {
  const { loading, data, ...rest } = useQueryApollo(query, {
    ...options,
  });

  return { data, loading, ...rest };
}

// export default useQuery;

