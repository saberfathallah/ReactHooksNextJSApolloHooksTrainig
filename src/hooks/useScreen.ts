import { useEffect, useState } from "react";
import { isMobile, isMobileWidth } from "@utils/isBrowser";
import { ESKTOP_BREAKPOINT } from "@constants/dimensions";

const useScreen = (isClient) => {
  const [screen, setScreen] = useState({
    screenWidth: ESKTOP_BREAKPOINT,
    isMobileSize: isMobile,
  });

  useEffect(() => {
    if (!isClient) {
      return;
    }

    // first load
    setScreen({
      screenWidth: window.innerWidth,
      isMobileSize: isMobileWidth(window.innerWidth),
    });

    const handleResize = () => {
      if (isClient) {
        setScreen({
          screenWidth: window.innerWidth,
          isMobileSize: isMobileWidth(window.innerWidth),
        });
      }
    };

    window.addEventListener("resize", handleResize);

    return () => window.removeEventListener("resize", handleResize);
  }, [isClient]);

  return screen;
};

export default useScreen;
