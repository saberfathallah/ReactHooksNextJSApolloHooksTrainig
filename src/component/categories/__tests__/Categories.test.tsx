import React from "react";
import renderer from "react-test-renderer";

import Categories from "../Categories";
import { useQuery } from "../../../hooks/useQuery";
import { useMutation } from "../../../hooks/useMutation";
import { CATEGORY } from '../../../../mock/categoriesMock';

jest.mock("../../../hooks/useQuery", () => ({
  useQuery: jest.fn(),
}));

jest.mock("../../../hooks/useMutation", () => ({
  useMutation: jest.fn(),
}));

// @ts-ignore
useQuery.mockImplementation(() => ({
  data: { getAllCategoriesQuery: { categories: [CATEGORY] } },
}));
// @ts-ignore
useMutation.mockImplementation(() => [() => {}, { loading: false }]);

describe("Categories", () => {
  it("should display posts snapshot", () => {
    const props = {
      selectCategoryId: () => '',
      changeCurrentCategory: () => '',
      isDisplayCategories: true,
      loading: false,
      data: {
        getAllCategoriesQuery: {
          categories: [CATEGORY],
        },
      },
    };
    const tree = renderer.create(<Categories {...props} />);
    expect(tree).toMatchSnapshot();
  });

  // it("should display loading snapshot", () => {
  //   const props = {
  //     selectCategoryId: () => '',
  //     changeCurrentCategory: () => '',
  //     isDisplayCategories: true,
  //     loading: true,
  //     data: undefined,
  //   };
    
  //   const tree = renderer.create(<Categories {...props} />);
  //   expect(tree).toMatchSnapshot();
  // });
});
