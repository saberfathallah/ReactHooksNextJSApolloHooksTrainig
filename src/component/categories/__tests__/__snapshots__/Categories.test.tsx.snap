// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories should display categories snapshot 1`] = `
<div>
  <ul
    aria-multiselectable={true}
    className="MuiTreeView-root makeStyles-root-1"
    role="tree"
  >
    <div>
      <li
        aria-expanded={false}
        aria-selected={false}
        className="MuiTreeItem-root"
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="treeitem"
        tabIndex={-1}
      >
        <div
          className="MuiTreeItem-content"
          onClick={[Function]}
          onMouseDown={[Function]}
        >
          <div
            className="MuiTreeItem-iconContainer"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </div>
          <div
            className="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
          >
            Sport
          </div>
        </div>
      </li>
    </div>
  </ul>
</div>
`;
